### PR TITLE
[APISIX] IMAP proxy

### PR DESCRIPTION
**Pull request by quantranhong1999**
[demo IMAP proxy.webm](https://user-images.githubusercontent.com/55171818/237060482-1547fcfd-f508-4bb6-8bca-3a47891bfbeb.webm)

Tested locally the IDLE time with telnet/openssl as well.

TODO: Investigate possible plugins for this stream proxy.

### CODE REVIEW

File: demo/apisix/conf/apisix.yaml [hunk 1]
```diff
@@ -164,6 +164,14 @@ routes:
       - OPTIONS
     plugin_config_id: jmap-plugin
 
+stream_routes:
+  - id: imap_starttls
+    server_port: 143
+    upstream_id: imap_starttls_upstream
+  - id: imap_ssl
+    server_port: 993
+    upstream_id: imap_ssl_upstream
```
Comment (chibenwa) [inline demo/apisix/conf/apisix.yaml:173]: Is it possible to do SSL termination on Apisix side?

**Review by chibenwa — COMMENTED** [review]
Do Apisix support the proxy protocol?

File: demo/apisix/conf/apisix.yaml (hunk 1 above)
Comment (quantranhong1999) [reply demo/apisix/conf/apisix.yaml:173]: Yes IMO:
https://apisix.apache.org/docs/apisix/stream-proxy/#accept-tls-over-tcp-connection

**Comment by quantranhong1999** [comment]
> Do Apisix support the proxy protocol?

APISIX is based on NGINX, and NGINX supports that. So maybe there is somewhere we can configure that.

File: demo/docker-compose.yml [hunk 2]
```diff
@@ -22,13 +22,17 @@ services:
...
       - "8001:80"
+      - "25:25"
     environment:
       - DOMAIN=tmail.com
     volumes:
       - ./tmail/jwt_publickey:/root/conf/jwt_publickey
       - ./tmail/jwt_privatekey:/root/conf/jwt_privatekey
       - ./tmail/jmap.properties:/root/conf/jmap.properties
       - ./tmail/usersrepository.xml:/root/conf/usersrepository.xml
+      - ./tmail/keystore:/root/conf/keystore
```
Comment (vttranlina) [inline demo/docker-compose.yml:33]:
```yaml
command:
   - --generate-keystore
```

then we don't need to mount the keystore file. A new feature in love :D

Comment (vttranlina) [reply demo/docker-compose.yml:33]: update! it is not yet in tmail

**Comment by vttranlina** [comment]
Can we test it after 30 minutes (or more) inactive Thunderbird, it still works

**Comment by quantranhong1999** [comment]
> Can we test it after 30 minutes (or more) inactive Thunderbird, it still works

I tested with telnet/openssl. For Thunderbird, IMO it is harder to see if after 30m Thunderbird just creates a new connection.

**Comment by Arsnael** [comment]
> Can we test it after 30 minutes (or more) inactive Thunderbird, it still works

You don't really know what TB is doing behind... he could be reconnecting. 

I would think login on imap via telnet is more appropriate, then you see how long it takes before being disconnected when idle ?

**Comment by vttranlina** [comment]
> I tested with telnet/openssl. For Thunderbird, IMO it is harder to see if after 30m Thunderbird just creates a new connection.
>I would think login on imap via telnet is more appropriate, then you see how long it takes before being disconnected when idle ?

yes, lost connection after idle/inactive is my concern, the telnet test is good.

**Comment by quantranhong1999** [comment]
PROXY protocol works with APISIX now:
```java
08:42:26.115 [INFO ] o.a.j.i.n.ImapChannelUpstreamHandler - Connection established from 172.22.0.7
08:42:26.146 [INFO ] o.a.j.i.n.HAProxyMessageHandler - Connection from 172.22.0.4 runs through 172.22.0.7 proxy
08:42:49.945 [INFO ] o.a.j.i.n.ImapChannelUpstreamHandler - Connection established from 172.22.0.7
08:42:49.948 [INFO ] o.a.j.i.n.HAProxyMessageHandler - Connection from 172.22.0.8 runs through 172.22.0.7 proxy
```

For plugins, only limited plugins are available for stream proxy type (TCP/UDP) and we likely get not much out of them: https://github.com/apache/apisix/tree/master/apisix/stream/plugins

- `limit-conn`: to limit number of concurrent requests. ~~ what we had in IMAP (`concurrentRequests`)
- `ip-restriction` (blacklist/whilelist IP): could bring limited value IMO
- Others: `mqtt-proxy`, `prometheus` (we had our Prometheus endpoint anyway), `syslog` are not helpful.

**Comment by chibenwa** [comment]
> limit-conn: to limit number of concurrent requests. ~~ what we had in IMAP (concurrentRequests)

With APISIX this should be distributed right?

With our existing set up, it's per james hosts.

**Comment by quantranhong1999** [comment]
> Can we run custom plugins for TCP?

Yes. But we need to write the plugin in Lua.
rf: https://apisix.apache.org/docs/apisix/plugin-develop/

For the plugin written in other languages e.g JAVA, I think it is only applied for L7 protocol here HTTP.

**Comment by chibenwa** [comment]
Closing as the benefits are not worth it: no additional features compared to HA Proxy.

Thanks for the experiment, that was interesting but we would likely not use this.

**Pull request closed without merging**